explanation/intro-to/ebpf.md

```diff
@@ -0,0 +1,122 @@
+(introduction-to-ebpf)=
+# Introduction to eBPF
+
+eBPF is a lightweight virtual machine within the kernel. This virtual machine
+is exposed to the user-space and is able to safely run sandboxed programs in
+the kernel-space.
+
+BPF was originally designed to perform network packet filtering. Thus its name:
+_Berkley Packet Filter_. Over time, BPF was extended to use more registers,
+support 64-bit registers, data stores (Maps), and more. Such enhancement
+allowed this extended version of BPF (eBPF) to be decoupled from the kernel
+networking subsystem - eBPF was transformed in a tool with a broader scope,
+capable of enhancing not only the networking experience, but also tracing,
+profiling, observability, security, etc. Given its broader range of
+applications, calling eBPF a packet filter no longer describes its full
+capabilities. eBPF is now considered to be a standalone term and no longer
+refers to _extended Berkley Packet Filter_, reflecting its expanded role. Also,
+the terms eBPF and BPF are used interchangeably, to refer to eBPF.
+
+User-space applications can load eBPF programs into the kernel as eBPF
+bytecode. Although you could write eBPF programs in bytecode, there are several
+tools which provide abstraction layers on top of eBPF so you do not need to
+write bytecode manually. These tools will then generate the bytecode which will
+in turn be loaded into the kernel.
+
+Once the eBPF program is loaded, it is then verified by the kernel before it
+can run. These checks include verifying if the eBPF program halts and will
+never get stuck in a loop, verifying that the program will not crash by
+checking the registers and stack state validity throughout the program, and
+ensuring the process loading the eBPF program has all the capabilities required
+by the eBPF program to run.
+
+After the verification step, the bytecode is Just-In-Time (JIT) compiled into
+machine-code to optimize the program's execution.
+
+eBPF is a powerful tool for server and system administrators. It provides the
```

Comment:
This feels a bit out of place after the "how it works" section, it's like we've gone back to the "who/what it's for" motivation, which should be near the top - I'd move this up to near the start (about line 5-6 I think?)

Reply:
(with a little rewriting to avoid duplicating words on the next paragraph)

Reply:
Here, I added a new "eBPF in Ubuntu Server" section (this was originally the beginning of a section). Does this help?

Reply:
The new title is good and suits the rest of the content, but I'd still prefer this first paragraph to be near the top (in the intro). It basically provides all the justification (the "why should I care?") that the reader is looking for to know whether it's relevant or interesting for them. Imo, once a reader is halfway down the page, they've already decided they're interested (and the "why" is a waste of their time at that point). If they don't get halfway down because we didn't provide a strong hook then they'll never see that this is relevant for them. This paragraph does a great job of explaining the "why", so I'd rather not bury it, but rather put it right in front of the reader's eyes :)